camelConnector/src/main/java/io/vantiq/extsrc/camelconn/discover/CamelDiscovery.java

```diff
@@ -98,22 +110,47 @@ Map<String, Set<String>> performComponentDiscovery(RouteBuilder rb, Properties p
     
             }
             Map<String, Set<String>> retVal = new HashMap<>();
+    
             retVal.put(SYSTEM_COMPONENTS, ecr.getSystemComponentsUsed());
             retVal.put(COMPONENTS_TO_LOAD, ecr.getComponentsToLoad());
             retVal.put(SYSTEM_DATAFORMATS, edfr.getSystemDataFormatsUsed());
             retVal.put(DATAFORMATS_TO_LOAD, edfr.getDataFormatsToLoad());
             return retVal;
-            // Note that ectx will be closed via the try-with-resources construct
         } catch (Exception e) {
             log.error("Trapped exception preparing or completing component discovery", e);
             throw new DiscoveryException("Exception preparing or completing component discovery processing", e);
+        } finally {
+            // The following is aimed at reducing the resource overhead that discovery may introduce.  The ctx.close()
+            // call claims to remove all stuff used, but it doesn't, leaving bits of expressions around that refer to
+            // the camel context under which they were built.  Consequently, we cannot close the context, so we'll do
+            // what we can to get rid of the detritus.
+    
+            try {
+                ctx.stopAllRoutes();
+            } catch (Exception e) {
+                log.error("Unable to stop all discovery routes:", e);
+            }
+            ctx.getRoutes().forEach(ctx::removeRoute);
+            ;
+            List<RouteDefinition> rds = new ArrayList<>(ctx.getRouteDefinitions());
+            rds.forEach( (rd) -> {
+                try {
+                    ctx.removeRouteDefinition(rd);
+                    log.debug("Removed discovery-time route definition: {}", rd.getShortName());
+                } catch (Exception e) {
+                    log.error("Failed to remove discovery-time route definition: {}", rd.getShortName());
+                    throw new RuntimeException(e);
+                }
+            });
+            ExpressionReifier.clearReifiers();
+            ProcessorReifier.clearReifiers();
         }
     }
     
     /**
      * Find the name of the camel component from which it runs these components based on the scheme.
      *
```

Comment:
Just removed the bad link -- IntelliJ notes it as a compiler error, so it's just confusing me late on Saturday nights.